x/projects/types/project.go

```diff
@@ -44,46 +44,53 @@ func ValidateProjectNameAndDescription(name string, description string) bool {
 	return true
 }
 
-func (project *Project) GetKey(projectKey string) ProjectKey {
-	for _, key := range project.ProjectKeys {
-		if key.Key == projectKey {
-			return key
-		}
-	}
-	return ProjectKey{}
+func NewProjectKey(key string) ProjectKey {
+	return ProjectKey{Key: key}
 }
 
-func (projectKey ProjectKey) IsKeyType(keyTypeToCheck ProjectKey_KEY_TYPE) bool {
-	for _, keytype := range projectKey.Types {
-		if keytype == keyTypeToCheck {
-			return true
-		}
-	}
-	return false
+func (projectKey ProjectKey) AddType(kind ProjectKey_Type) ProjectKey {
+	projectKey.Kinds |= uint32(kind)
+	return projectKey
+}
+
+func ProjectAdminKey(key string) ProjectKey {
+	return NewProjectKey(key).AddType(ProjectKey_ADMIN)
+}
+
+func ProjectDeveloperKey(key string) ProjectKey {
+	return NewProjectKey(key).AddType(ProjectKey_DEVELOPER)
 }
 
-func (projectKey *ProjectKey) AppendKeyType(typesToAdd []ProjectKey_KEY_TYPE) {
-	for _, keytype := range typesToAdd {
-		if !projectKey.IsKeyType(keytype) {
-			projectKey.Types = append(projectKey.Types, keytype)
+func (projectKey ProjectKey) IsType(kind ProjectKey_Type) bool {
+	return projectKey.Kinds&uint32(kind) != 0x0
+}
+
+func (projectKey ProjectKey) IsTypeValid() bool {
+	const keyKindsAll = (uint32(ProjectKey_ADMIN) | uint32(ProjectKey_DEVELOPER))
+
+	return projectKey.Kinds != uint32(ProjectKey_NONE) &&
```

Comment:
because it doesn't do the job. we want to return true if :
1. it has only the allowed types:  `kinds & ~keyKindsAll == 0x0`, and 
2. it has at least one type:  `kinds != 0x0`
your suggestion tests whether it has a disallowed flag (i.e. the opposite of 1) but ignores the second test.

Reply:
projectKey.Kinds & keyKindsAll != 0x0 ensures both conditions